### PR TITLE
Light Class

### PR DESCRIPTION
**Pull request by allison324**
We made Light class and mainly copied everything from last year's code(same functions and everything) - the only things that we changed are the colors and the enums. Also added a const int for the canid in RobotMap.java.

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: probably add a HOLDING_NOTE enum

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: I would reorder these so that OFF and PARTY_MODE are at the end. this is nitpicky though

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: I could be wrong, but shouldn't OFF just be no lights are on?

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: 1) ALIGNING_TO_TAG is in here twice --> copy-paste typo?
2) I think that the first ALIGNING_TO_TAG is supposed to be READY_TO_SHOOT because you say that is the green color earlier

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: i don't know if this is necessary, but usually we make variables for each value so they aren't unlabeled. so maybe make variables in LightsConstatnts.java for the rest of these parameters (1, 0.5)?

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java:60]: I know you were looking at the minibot Lights class for inspiration. The reason why we needed this second if-statement in the minibot was because NO_TAG was a blinking strobe animation (like how PARTY_MODE is an animation).

if NO_TAG is just a solid color like the other LightCode options (which is what you have commented and in the tree map) then you don't need a separate if statement and function to change the lights to the NO_TAG colorway

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: this isn't needed if you're making NO_TAG a solid color (which is what you have defined in the tree map). We can have a real discussion about the colors/patterns we want to use later on though.

File: src/main/java/frc/robot/RobotMap.java
Comment (avakimc) [inline src/main/java/frc/robot/RobotMap.java]: this is nitpicky but the comment is kind of misleading, in my opinion. either delete it or edit

**Review by avakimc — CHANGES_REQUESTED** [review]
i have requested some changes for this class

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: this comment is not accurate as lightOptionsMap is not an array. also i would use the /** comment here */ format for the comment so that hovering over the variable lightOptionsMap will display the comment. make sure to put that comment on the line where you say private TreeMap<LightCode, Integer[]> lightOptionsMap; (line 14) though.

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java]: if OFF is not 0,0,0 the comment shouldn't be "grey" right? shouldn't it be like, "no lights are on" or something?

File: src/main/java/frc/robot/subsystems/lights/LightsConstants.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/LightsConstants.java]: I could be wrong here but I looked over the documentation and "brightness" is also a double. therefore, LIGHT_BRIGHTNESS should be a double with a value of 1.0. I think that having it as an int also works but generally we like to make it a double and add the .0 to be more clear
Screenshot 2024-02-12 at 7 20 50 PM

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java:72]: you still have NO_TAG as a blinking animation and as a solid color (from the treemap). you shouldn't / can't have both

**Review by avakimc — CHANGES_REQUESTED** [review]
added some more comments

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (tails618) [reply src/main/java/frc/robot/subsystems/lights/Lights.java:35]: Agree that we want the comment to display when hovering, but that means the comment must be above where we *create* the variable, on line 14.

File: src/main/java/frc/robot/subsystems/lights/LightsConstants.java
Comment (tails618) [reply src/main/java/frc/robot/subsystems/lights/LightsConstants.java]: Yes, brightness should be a double for animations (it's [0,1], so if it was an int it wouldn't really be brightness and it would just be on/off). An integer will work (Java can typecast an integer n to a double n.0) but it lacks clarity and there really isn't a reason to have it as an int here so this should be a double.

File: src/main/java/frc/robot/subsystems/lights/Lights.java
Comment (avakimc) [inline src/main/java/frc/robot/subsystems/lights/Lights.java:59]: you deleted NO_TAG from the treemap, which was good. you also deleted it as a LightCode but you still need NO_TAG as an enum option (that way you can set it as a desired color and stuff). so i don't think this will build

**Review by avakimc — CHANGES_REQUESTED** [review]
last requested change

**Review by avakimc — APPROVED** [review]
approved. probably don't merge until basel or asher also approve

**Review by tails618 — APPROVED** [review]

**Pull request merged**
